rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD TBD - Headless Authentication
+
+## Required Approvers
+
+* Engineering: @r0mant && @jakule
+* Security: @reed || @jentfoo
+* Product: @xinding33 || @klizhentas
+
+## What
+
+Support headless authentication for Teleport clients.
+
+Headless authentication is a non-interactive form of authentication that relies on a secondary party to authenticate a user. In our case, headless authentication will be used to authenticate a user on a remote machine using their credentials and MFA device on their local machine.
+
+## Why
+
+Some users have valid use cases for making `tsh` requests on a remote, sometimes shared, machine.
+
+For example:
+
+ 1. Connecting to Teleport Nodes.
+ 1. Performing `tsh scp` from one Teleport Node to another Teleport Node.
+
+However, `tsh` does not currently work well on remote machines for two reasons:
+
+ 1. Client credentials are usually saved to disk, where they may be overwritten, used, or stolen by anyone with access to the remote machine.
+ 1. Teleport widely uses MFA to authenticate client logins and actions, which may require direct access to a WebAuthn device. For example, with Per-session MFA, each SSH connection to a Teleport Node may require MFA verification.
+
+Headless authentication will provide a secure way to support these types of remote use cases.
+
+## Details
+
+### Security
```

Comment:
Headless logins are based on an unauthenticated endpoint that writes backend data against a given user. I think we should include here:

* What measures are we taking against the new endpoint being used to flood / DoS Teleport?

* What measures are we taking against a malicious actor flooding users with headless requests, in the hopes that they get tired and approve some?

Reply:
> * What measures are we taking against the new endpoint being used to flood / DoS Teleport?

I'll update the RFD to specify that we will perform rate limiting, and write the `HeadlessAuthentication` to the backend with a 1 minute TTL to prevent overflow. Do you think this is sufficient?

> * What measures are we taking against a malicious actor flooding users with headless requests, in the hopes that they get tired and approve some?

The user would not be notified of headless requests, so these attempts should disappear into the void after 1 minute. 

If they do go to the page or enter the headless SSO link, they will be warned along the lines of "You should never approve a headless login request that you did not initiate." We need to express the importance of this, in the same way that support platforms often say "We will never ask you for your password."

Do you think we need an additional counter measure?

Reply:
Sounds good to me on both accounts.  👍